spyx/nir.py

```diff
@@ -0,0 +1,111 @@
+import nir
+
+
+def _nir_node_to_spyx_node(node: nir.NIRNode):
+    """Converts a NIR node to a Spyx node."""
+    # NOTE: all nodes have node.input_type and node.output_type
+    # which specify the input and output shape of the node.
+    if isinstance(node, (nir.Input, nir.Output)):
+        node.input_type
+        return None
+
+    elif isinstance(node, nir.Affine):
+        # NOTE: node.weight, node.bias are npy arrays
+        pass
+
+    elif isinstance(node, nir.Linear):
+        # NOTE: node.weight
+        pass
+
+    elif isinstance(node, nir.Conv1d):  # not needed atm
+        # NOTE: node.bias, node.weight
+        # node.dilation, node.groups, node.padding, node.stride
+        pass
+
+    elif isinstance(node, nir.Conv2d):
+        # NOTE: node.bias, node.weight
+        # node.dilation, node.groups, node.padding, node.stride
+        pass
+
+    elif isinstance(node, nir.IF):
+        # NOTE: node.r, node.v_threshold
+        pass
+
+    elif isinstance(node, nir.LIF):
+        # NOTE: node.r, node.v_threshold, node.tau, node.v_leak
+        pass
+
+    elif isinstance(node, nir.CubaLIF):
+        # NOTE: node.r, node.v_threshold, node.v_leak
+        # node.tau_mem, node.tau_syn
+        # node.w_in
+        pass
+
+    elif isinstance(node, nir.Flatten):
+        # NOTE: node.start_dim, node.end_dim
+        pass
+
+    elif isinstance(node, nir.I):  # not needed atm
+        pass
+
+    elif isinstance(node, nir.Sequence):  # not needed atm
+        pass
+
+    elif isinstance(node, nir.Scale):  # not needed atm
+        pass
+
+    elif isinstance(node, nir.Delay):  # not needed atm
+        pass
+
+    elif isinstance(node, nir.Threshold):  # not needed atm
+        pass
+
+
+def to_nir(spyx_pytree, input_shape, output_shape) -> nir.NIRGraph:
+    """Converts a Spyx network to a NIR graph."""
+    # construct the edge list for the NIRGraph
+    keys = list(spyx_pytree.keys())
+    edges = [(keys[i], keys[i + 1]) for i in range(len(keys) - 1)] # assume linear connectivity
```

Comment:
will the keys of the pytree always be ordered? if not, the edges of the NIR graph would be incorrect, wouldn't they?